File: src/main/java/at/jku/ssw/coco/intellij/CocoUtil.java
package at.jku.ssw.coco.intellij;

import at.jku.ssw.coco.intellij.psi.*;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiManager;
import com.intellij.psi.search.FileTypeIndex;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.util.indexing.FileBasedIndex;
import org.apache.commons.lang.StringUtils;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;
import java.util.stream.Collectors;

/**
 * @author Thomas Scheinecker <a href="mailto:tscheinecker@gmail.com">tscheinecker@gmail.com</a>
 */
public class CocoUtil {

    @NotNull
    public static List<String> findCompilerNames(@Nullable Project project) {
        return findCompilers(project)
                .stream()
                .map(CocoCompiler::getName)
                .filter(it -> !Objects.isNull(it))
                .collect(Collectors.toList());
    }

    @Contract("_, null -> null")
    @Nullable
    public static CocoCompiler findCompiler(@NotNull PsiFile file, @Nullable String name) {
        return findByName(findCompilers(file), name);
    }

    @NotNull
    public static List<CocoCompiler> findCompilers(@NotNull PsiFile file) {
        return findCompilers(file.getProject())
                .stream()
                .filter(compiler -> file.equals(compiler.getContainingFile()))
                .collect(Collectors.toList());
    }

    public static List<CocoCompiler> findCompilers(@Nullable Project project, @NotNull String name) {
        return findCompilers(project)
                .stream()
                .filter(compiler -> name.equals(compiler.getName()))
                .collect(Collectors.toList());
    }

    @NotNull
    public static List<CocoCompiler> findCompilers(@Nullable Project project) {
        if (project == null) {
            return Collections.emptyList();
        }

        Collection<VirtualFile> virtualFiles = FileBasedIndex.getInstance().getContainingFiles(FileTypeIndex.NAME, CocoFileType.INSTANCE, GlobalSearchScope.allScope(project));
        List<CocoCompiler> compilers = new ArrayList<>();
        for (VirtualFile virtualFile : virtualFiles) {
            CocoFile cocoFile = (CocoFile) PsiManager.getInstance(project).findFile(virtualFile);
            if (cocoFile != null) {
                CocoCompiler[] cocoCompilers = PsiTreeUtil.getChildrenOfType(cocoFile, CocoCompiler.class);
                if (cocoCompilers != null) {
                    compilers.addAll(Arrays.asList(cocoCompilers));
                }
            }
        }
        return compilers;
    }

    @NotNull
    public static List<CocoSetDecl> findCharacterDeclarations(@NotNull PsiFile file) {
        CocoScannerSpecification scannerSpecification = PsiTreeUtil.getChildOfType(file, CocoScannerSpecification.class);

        if (scannerSpecification == null) {
            return Collections.emptyList();
        }

        CocoCharacters characters = scannerSpecification.getCharacters();

        if (characters == null) {
            return Collections.emptyList();
        }

        return characters.getSetDeclList();
    }

    @Contract("_, null -> null")
    @Nullable
    public static CocoSetDecl findCharacterDeclaration(@NotNull PsiFile file, @Nullable String name) {
        return findByName(findCharacterDeclarations(file), name);
    }

    @NotNull
    public static List<CocoProduction> findProductions(@NotNull PsiFile file) {
        CocoParserSpecification parserSpecification = PsiTreeUtil.getChildOfType(file, CocoParserSpecification.class);

        if (parserSpecification == null) {
            return Collections.emptyList();
        }

        return parserSpecification.getProductionList();
    }

    @Contract("_, null -> null")
    @Nullable
    public static CocoProduction findProduction(@NotNull PsiFile file, @Nullable String name) {
        return findByName(findProductions(file), name);
    }

    @NotNull
    public static List<CocoTokenDecl> findTokenDecls(@NotNull PsiFile file) {
        CocoScannerSpecification scannerSpecification = PsiTreeUtil.getChildOfType(file, CocoScannerSpecification.class);

        if (scannerSpecification == null) {
            return Collections.emptyList();
        }

        CocoTokens tokens = scannerSpecification.getTokens();
        if (tokens == null) {
            return Collections.emptyList();
        }

        return tokens.getTokenDeclList();
    }

    @Contract("_, null -> null")
    @Nullable
    public static CocoTokenDecl findTokenDecl(@NotNull PsiFile file, @Nullable String name) {
        return findByName(findTokenDecls(file), name);
    }

    @Contract("_, null -> null")
    @Nullable
    private static <T extends CocoNamedElement> T findByName(@NotNull Collection<T> collection, @Nullable String name) {
        if (StringUtils.isBlank(name)) {
            return null;
        }

        Optional<T> first = collection
                .stream()
                .filter(item -> name.equals(item.getName()))
                .findFirst();

        return first.isPresent() ? first.get() : null;
    }
}

File: src/main/kotlin/me/salzinger/intellij/coco/CocoUtil.kt
@file:Suppress("TooManyFunctions")

package me.salzinger.intellij.coco

import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.PsiManager
import com.intellij.psi.PsiNameIdentifierOwner
import com.intellij.psi.search.FileTypeIndex
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.util.PsiTreeUtil
import me.salzinger.intellij.coco.psi.CocoCocoInjectorHost
import me.salzinger.intellij.coco.psi.CocoCompiler
import me.salzinger.intellij.coco.psi.CocoGlobalFieldsAndMethods
import me.salzinger.intellij.coco.psi.CocoNamedElement
import me.salzinger.intellij.coco.psi.CocoParserSpecification
import me.salzinger.intellij.coco.psi.CocoPragmaDecl
import me.salzinger.intellij.coco.psi.CocoProduction
import me.salzinger.intellij.coco.psi.CocoScannerSpecification
import me.salzinger.intellij.coco.psi.CocoSetDecl
import me.salzinger.intellij.coco.psi.CocoTokenDecl

fun findCompilers(file: PsiFile): List<CocoCompiler> {
    return PsiTreeUtil.findChildrenOfType(file, CocoCompiler::class.java).toList()
}

fun findParserSpecification(file: PsiFile): CocoParserSpecification? {
    return PsiTreeUtil.getChildOfType(
        file,
        CocoCocoInjectorHost::class.java
    )?.parserSpecification
}

fun findTokenDecls(file: PsiFile): List<CocoTokenDecl> {
    return findScannerSpecification(file)?.tokens?.tokenDeclList ?: emptyList()
}

fun findPragmaDecls(file: PsiFile): List<CocoPragmaDecl> {
    return findScannerSpecification(file)?.pragmas?.pragmaDeclList ?: emptyList()
}

fun findNearestCocoNamedElement(element: PsiElement): CocoNamedElement {
    var searchElement: PsiElement? = element
    while (searchElement != null) {
        if (searchElement is CocoNamedElement) {
            return searchElement
        }
        searchElement = searchElement.parent
    }

    throw NullPointerException("Couldn't find CocoNamedElement for element $element")
}

fun findTokenDecls(project: Project?): List<CocoTokenDecl> {
    return getAllFiles(project)
        .flatMap { findTokenDecls(it) }
}

fun findCharacterDecls(project: Project?): List<CocoSetDecl> {
    return getAllFiles(project)
        .flatMap { findCharacterDeclarations(it) }
}

fun findProductions(project: Project?): List<CocoProduction> {
    return getAllFiles(project)
        .flatMap { findProductions(it) }
}

fun findGlobalFieldsAndMethods(file: PsiFile): CocoGlobalFieldsAndMethods? {
    return PsiTreeUtil.getChildOfType(
        file,
        CocoCocoInjectorHost::class.java
    )?.globalFieldsAndMethods
}

fun findScannerSpecification(file: PsiFile): CocoScannerSpecification? {
    return PsiTreeUtil.getChildOfType(
        file,
        CocoCocoInjectorHost::class.java
    )?.scannerSpecification
}

fun findProductions(file: PsiFile): List<CocoProduction> {
    return findParserSpecification(file)?.productionList ?: emptyList()
}

fun findCharacterDeclarations(file: PsiFile): List<CocoSetDecl> {
    return findScannerSpecification(file)?.characters?.setDeclList ?: emptyList()
}

fun <T : CocoNamedElement> List<T>.findByName(name: String?): T? {
    if (name.isNullOrBlank()) {
        return null
    }

    return filterByName(name).firstOrNull()
}

fun <T : PsiNameIdentifierOwner> List<T>.filterByName(name: String?): List<T> {
    return if (name.isNullOrBlank()) {
        this
    } else {
        filter { item -> name == item.name }
    }
}

private fun getAllFiles(project: Project?): List<PsiFile> {
    if (project == null) {
        return emptyList()
    }

    val psiManager = PsiManager.getInstance(project)

    return FileTypeIndex.getFiles(CocoFileType.INSTANCE, GlobalSearchScope.allScope(project))
        .mapNotNull(psiManager::findFile)
}
